### PR TITLE
Feature: Extract GUIScript doc-strings through Python interpreter.

### PR DESCRIPTION
**Pull request by kindaro**
## Description

Solves https://github.com/gemrb/gemrb.github.io/issues/9.

There is a build target added to CMake that builds an executable `run_script` with the obvious function. It initializes as little of the game as was needed for the goal at hand.

#### Directions for further study:

* How many of the GUIScript functions this interpreter is initialized enough to actually run?
* A small nuisance is that the file `include.py` may not be found, resulting in a non-critical error. Possibly we may want to make the importing of this file optional.
* Recent Pandoc — [at least version 2.6][changelog] — is required to convert from the DokuWiki format.
   + I am not sure why it escapes apostrophes, but it does not affect the HTML output.
* I am going to squash and rebase when this branch is ready to be merged.

[changelog]: https://github.com/jgm/pandoc/blob/master/changelog.md#pandoc-26-2019-01-30

## Checklist

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] I used the same coding style as the surrounding code 
- [ ] I have tested the proposed changes
- [ ] I extended the documentation, if necessary

### CODE REVIEW

**Comment by lynxlynxlynx** [comment]
Please first:
- make it use the style already used in the project, including the headers
- fix the build errors
- include docs on how to use the python runner + in the script itself
- then there's several things from the bug that it's not doing yet

**Comment by kindaro** [comment]
### Let me see if I understand you well.

* [ ] Style.

  > make it use the style already used in the project, including the headers

  * [ ] Use `clang-format`.
  * [ ] Use `autopep8`.
  * [ ] Add license headers.
  * [ ] Sort `include` directives as described in https://github.com/gemrb/gemrb/blob/master/gemrb/docs/en/CodingStyle.txt.

* [ ] Build errors.

  > fix the build errors

  * [ ] On Windows.
  * [ ] On Linux.

* [ ] Document.

  > include docs on how to use the python runner + in the script itself

* [ ] Add missing functionality.

  > then there's several things from the bug that it's not doing yet

  * [ ] Add links to index, as in https://github.com/gemrb/gemrb/blob/master/admin/guidoc_wikifier.sh#L39-L42:

            # append a link to the function index
            for txt_file in "$out_dir"/*.txt; do
              echo -e '\n[[guiscript:index|Function index]]' >> "$txt_file"
            done

  * [ ] Add front matter, as described in https://github.com/gemrb/gemrb.github.io/issues/9#issuecomment-612354143:

            ---
            title: blablabla
            ---

  * Add index? What kind of index?

### Sounds about right?

I shall look into it tomorrow.

**Comment by lynxlynxlynx** [comment]
style: no, take it literally if that helps.
functionality: frontmatter (should also include a module key), file naming (.md)

**Comment by kindaro** [comment]
@lynxlynxlynx  I cannot quite understand you. Can you spell it all out in more detail?

* Style. I have no idea what you are talking about. What is _«it»_, what does _«no»_ negate?
* Functionality. How should the front matter look? What is a module key?

Note: There is example usage in the commit message of the last commit. It is a simple loop. I suppose I can put it right into the Python file.

**Comment by lynxlynxlynx** [comment]
"no" negates your statement, while "it" refers to mine. Your subtasks are going in the wrong direction of the bat.

frontmatter: like quoted, just add a line with "module: name-of-the-module" as well.

File: admin/extract_docstrings.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+def adjust_document(document, **kwargs):
+	# Pass through Pandoc.
+	# Add an appropriately filled module field.
+	# Add the link to the function index.
+	metadata = ["--metadata={}:{}".format(kwname, kwvalue)
+		for kwname, kwvalue in kwargs.items()]
+	pandoc_invocation = ["pandoc", "--from=dokuwiki",
+			"--to=markdown+yaml_metadata_block", "--standalone"] + metadata
+	pipe = Popen(pandoc_invocation, stdin=PIPE, stdout=PIPE)
```
Comment (lynxlynxlynx) [inline admin/extract_docstrings.py:34]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=5396854)

**Comment by lynxlynxlynx** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=5396854&bid=17690860)

File: admin/extract_docstrings.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+import GemRB
+
+import inspect
+import tempfile
+import os
+from subprocess import Popen, PIPE
```
Comment (lynxlynxlynx) [inline admin/extract_docstrings.py:23]: Codacy Issue found: [Consider possible security implications associated with Popen module.](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=5396854)

File: admin/extract_docstrings.py (hunk 1 above)
Comment (kindaro) [reply admin/extract_docstrings.py:34]: This is not relevant.
* Only sane data is given as arguments.
* [The _shell_ option is _False_ by default.](https://docs.python.org/2.7/library/subprocess.html#subprocess.Popen)

File: admin/extract_docstrings.py (hunk 2 above)
Comment (kindaro) [reply admin/extract_docstrings.py:23]: _Popen_ is used in one place, and the invocation appears to be safe.

**Comment by lynxlynxlynx** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=5396854&bid=17690860)

File: admin/extract_docstrings.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+def adjust_document(document, **kwargs):
+	# Pass through Pandoc.
+	# Add an appropriately filled module field.
+	# Add the link to the function index.
+	metadata = ["--metadata={}:{}".format(kwname, kwvalue)
+		for kwname, kwvalue in kwargs.items()]
+	pandoc_invocation = ["pandoc", "--from=dokuwiki",
+			"--to=markdown+yaml_metadata_block", "--standalone"] + metadata
+	pipe = Popen(pandoc_invocation, stdin=PIPE, stdout=PIPE)
```
Comment (lynxlynxlynx) [inline admin/extract_docstrings.py:34]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=5396854)

File: admin/extract_docstrings.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+import GemRB
+
+import inspect
+import tempfile
+import os
+from subprocess import Popen, PIPE
```
Comment (lynxlynxlynx) [inline admin/extract_docstrings.py:23]: Codacy Issue found: [Consider possible security implications associated with Popen module.](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=5396854)

**Comment by kindaro** [comment]
@lynxlynxlynx  Now that we have the Pandoc framework in place, I wonder if there is a more structured way to arrive to the end result. For example, we may provide a custom template that the document would be embedded in. In particular, the link to the function index may belong to the template — it looks quite out of place hard-wired. What do you think? The possibilities are endless. We can also define custom filters, say, to turn bold text into headers.

**Comment by lynxlynxlynx** [comment]
Nah, we need to fix the source eventually and things like automatically adding the "go back" index link are more cleanly done via jekyll layouts.

**Comment by kindaro** [comment]
I am not interested in finishing up this work anymore, since you [_«stole the kill»_][stole] in https://github.com/gemrb/gemrb.github.io/issues/9 . _(Note that my solution was already working by the time, only needed some build system tuning.)_ Ugly end to our brief relationship.

You have a great project. Best of luck. I hope you do better with future prospective contributors.

[stole]: https://tvtropes.org/pmwiki/pmwiki.php/Main/KillSteal

**Comment by lynxlynxlynx** [comment]
A few more changes would be needed to make it usable (other module, title, index, maybe more), but the overall design was now clear enough to be fully evaluated. And as I feared and warned from the start, it failed it.

**Comment by kindaro** [comment]
@lynxlynxlynx  So tell me, what exactly did my design fail at?

Speaking of features — you never clearly specified your requirements in detail, but:
* I did my best to extract them from the body of the relevant commentary.
* I put them in [the check list above][checklist].
* I explicitly solicited your feedback and integrated it into the check list.

Then I programmed every entry on that check list. If it is not usable despite having all the features you asked for, then you asked for wrong features.

But now you are saying the design is not a match to that of your solution. So tell me, what is it that my design is not good enough for? Can you put forward a technical argument? Please, help me learn from my mistakes.

[checklist]: https://github.com/gemrb/gemrb/pull/698#issuecomment-617407311

**Comment by lynxlynxlynx** [comment]
I don't know why you decided to ignore the other module, but that would have been easy to amend. Same for the other things I mentioned — not all of them were clear before we had the final theme or some test content up.

The design is needlessly complex and even affects the build system. Just not a maintenance burden worthy of accepting, since it is avoidable. I asked for a jeweller's hammer and you ended up making a mallet (and some tanks on the way).

On a simplistically technical level I don't have many reservations. What Codacity reported is not practically important; neither windows users often having broken python (no modules; but they don't have gawk either) — though I don't know about pandoc, since you didn't use python bindings; the extra dependency is annoying, but like stated before, not critical, even for buildbot integratoin.

**Comment by kindaro** [comment]
@lynxlynxlynx  I see, so it amounts to an evaluation of the technical debt incurred by either solution.

By my book, applying regular expressions to languages that are not regular is a maintenance burden, and [I put forward an argument to this point][argument] at the outset. Your blessing followed, so I had an impression that my effort will be welcome. Had I inferred that my appeals to correctness are lost on you and that my further work holds no added value by your measure _(which is quite obvious from the tone of your messages, in hindsight)_, all the drama would have been avoided. Silly me to be so blind.

[argument]: https://github.com/gemrb/gemrb.github.io/issues/9#issuecomment-612372965

**Comment by lynxlynxlynx** [comment]
You put forth arguments and ignore counterarguments, so there's nothing really to add. But design is not a hard science and I'm just one limited guy, so it would be both silly and arrogant to shoot things down at first sign of different thinking or trouble. Once you made it far enough for a proper evaluation, that's what you got, and the final simplifications just weren't enough.

**Comment by kindaro** [comment]
@lynxlynxlynx  It is true that I did not give an answer to some of your claims. The reason is that I sought to avoid appearing hostile. [_(Since that is how you took my argument when I allowed myself to put it forward)_][hostile] Your line of reasoning revolves around the proposition that a shell script with regular expressions is sufficient because it can be made to work for the doc-strings as they currently stand. However, that is the definition of unnecessary interdependence, therefore poor design. There are many more weak points in your messages, and we can walk through them if you are interested. Having revisited our previous correspondence, I can see now that abstract considerations of design quality have little value in your eyes, and you are of course entitled to that — I only wish I saw that earlier.

[hostile]: https://github.com/gemrb/gemrb.github.io/issues/9#issuecomment-612382810

**Pull request closed without merging**